### PR TITLE
add -Wnarrowing as error to cmake builds

### PR DESCRIPTION
**Pull request by dagitses**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89207

### CODE REVIEW

**Review by wconstab — APPROVED** [review]

**Review by malfet — APPROVED** [review]
LGTM, but curious why this needs to be limited to just clang builds?

File: CMakeLists.txt [hunk 1]
```diff
@@ -837,6 +837,7 @@ if(NOT MSVC)
   append_cxx_flag_if_supported("-Wno-error=deprecated-declarations" CMAKE_CXX_FLAGS)
   append_cxx_flag_if_supported("-Wvla-extension" CMAKE_CXX_FLAGS)
   if("${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang")
+    string(APPEND CMAKE_CXX_FLAGS " -Werror=c++11-narrowing")
```
[the remark this answers is not in the document]
Comment (dagitses) [reply CMakeLists.txt]: `-Wc++11-narrowing` doesn't exist in GCC, but `-Wnarrowing` exists in both and seems to catch everything.

Made all changes.

**Pull request closed without merging**